// cote/dist/lib/Component.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var EventEmitter = require('eventemitter2').EventEmitter2;
var Discovery = require('./Discovery');

module.exports = function (_EventEmitter) {
    _inherits(Component, _EventEmitter);

    function Component(advertisement, discoveryOptions) {
        _classCallCheck(this, Component);

        var _this = _possibleConstructorReturn(this, (Component.__proto__ || Object.getPrototypeOf(Component)).call(this, {
            wildcard: true, // should the event emitter use wildcards.
            delimiter: '::', // the delimiter used to segment namespaces, defaults to `.`.
            newListener: false, // if you want to emit the newListener event set to true.
            maxListeners: 2000 // the max number of listeners that can be assigned to an event, defaults to 10.
        }));

        advertisement.key = _this.constructor.environment + (advertisement.key || '');

        _this.advertisement = advertisement;
        _this.advertisement.axon_type = _this.type;

        _this.discoveryOptions = discoveryOptions || {};
        _this.discovery = Discovery(_this.advertisement, _this.discoveryOptions);

        _this.discovery.on('added', _this.onAdded.bind(_this));
        _this.discovery.on('removed', _this.onRemoved.bind(_this));
        return _this;
    }

    _createClass(Component, [{
        key: 'close',
        value: function close() {
            if (!this.discovery) return;

            this.discovery.stop();

            this.discovery.broadcast && this.discovery.broadcast.socket && this.discovery.broadcast.socket.close();
        }
    }, {
        key: 'onAdded',
        value: function onAdded(obj) {
            if (obj.advertisement.axon_type != this.oppo) return;
            if (obj.advertisement.key != this.advertisement.key) return;
            if (this.advertisement.namespace != obj.advertisement.namespace) return;

            this.emit('added', obj);
        }
    }, {
        key: 'onRemoved',
        value: function onRemoved(obj) {
            if (obj.advertisement.axon_type != this.oppo) return;
            if (obj.advertisement.key != this.advertisement.key) return;
            if (this.advertisement.namespace != obj.advertisement.namespace) return;

            obj && obj.sock && obj.sock.close();
            this.emit('removed', obj);
        }
    }]);

    return Component;
}(EventEmitter);
